### DISCUSSION TITLE
Ghcr.io docker HTTP API

### DISCUSSION THREAD

**Discussion by MJRichardson** [community/community#26299] [Packages]
<p>Hi,</p>
<p>Does the GitHub container registry (<a href="http://ghcr.io" rel="nofollow noopener">ghcr.io</a>) support the docker V2 HTTP API’s?</p>
<p>Specifically, does it support listing repositories and tags (<a href="https://docs.docker.com/registry/spec/api/#listing-repositories" rel="nofollow noopener">https://docs.docker.com/registry/spec/api/#listing-repositories</a>)?</p>

**Comment by christian-korneck** [comment]
<p>I guess the answer is “yes”. I’m assuming this from:</p>
<p>I can list the tags of a <a href="http://ghcr.io" rel="nofollow noopener">ghcr.io</a> repo with <a href="https://github.com/containers/skopeo" rel="nofollow noopener">skopeo</a>:</p>
<pre><code class="lang-auto">docker run --rm quay.io/skopeo/stable list-tags docker://ghcr.io/my-github-user/hello-world
{
    "Repository": "ghcr.io/my-github-user/hello-world",
    "Tags": [
        "latest"
    ]
}
</code></pre>
<p>and when I run it with <code>--debug</code> there are clues that it’s indeed using the <code>v2</code> endpoint:</p>
<pre><code class="lang-auto">level=debug msg="GET https://ghcr.io/v2/my-github-user/hello-world/manifests/latest"
</code></pre>

**Comment by clarkbw** [comment, marked as answer]
<p><code>/v2/_catalog</code> support is in our backlog! It hasn’t arrived yet but we know we need it.</p>

**Comment by MJRichardson** [comment]
<p>Thanks for the confirmation.</p>
<p>For context (and just in case it helps nudge it up the priority list), we make <a href="https://octopus.com/" rel="nofollow noopener">Octopus Deploy</a>, and when integrating with container registries we use the HTTP catalog API to list the available repositories and tags in the UI.</p>

**Comment by clarkbw** [comment]
<p>Thanks for that context! It does help.</p>
<p>Are there any other features / APIs that would be useful?</p>

**Comment by jgamblin** [comment]
<p>I know I am not the OP but here is what the DockerHub API provides now and it would be useful to have parity.</p>
<p><a href="https://hub.docker.com/v2/repositories/jgamblin/tiny-tor/" class="onebox" target="_blank" rel="nofollow noopener">https://hub.docker.com/v2/repositories/jgamblin/tiny-tor/</a></p>

**Comment by nohaapav** [comment]
<p>I guess it will be much more easy to publish graphql v4 api to list all repositories in registry related to user or organization than enabling <code>/_catalog</code> api which list basically all images. But both can do the job here.</p>

**Comment by Nicceboy** [comment]
<p>It would be really useful to modify README and package descriptions through API. Not every package could be linked to single GitHub repository, or the repository might have multiple packages.</p>
<p>Also, if the implementation of repository listing shows last modified time or Manifest hash, that would be neat - based on the list I could see which packages have changed since querying API last time.</p>

**Comment by minor-fixes** [comment]
<p>Is there a way to list all ghcr containers in an organization?</p>
<p>We would use the _catalog endpoint, but as mentioned above it is not implemented.</p>
<p>GraphQL can be used to list images on the old Docker packages, but this does not include ghcr images as per <a href="https://github.community/t/github-container-registry-ghcr-io-packages-not-appearing-in-api/130133">this question</a>.</p>
<p>The <a href="https://docs.github.com/en/rest/reference/packages" rel="noopener nofollow ugc">Packages REST API</a> can supposedly be used to interact with ghcr images, but only if you know the image name already!</p>
<p>How can one programmatically list all ghcr image names in an org?</p>

**Comment by lioskaderon** [comment]
<p><a class="mention" href="/clarkbw">@clarkbw</a> is there any update regarding the v2 support?</p>
<p>Thanks</p>

**Comment by gumpt** [comment]
<p>Have the same question, would love to know if/when this is planned.</p>

**Comment by elasticdotventures** [comment]
<p><a class="mention" href="/clarkbw">@clarkbw</a> FYI - this is blocking vscode issue <span class="hashtag">#869</span></p>
<aside class="onebox githubissue" data-onebox-src="https://github.com/microsoft/vscode-docker/issues/869">
  <header class="source">

      <a href="https://github.com/microsoft/vscode-docker/issues/869" target="_blank" rel="noopener nofollow ugc">github.com/microsoft/vscode-docker</a>
  </header>

  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Issue">
	  <svg width="60" height="60" class="github-icon" viewbox="0 0 14 16" aria-hidden="true"><path d="M7 2.3c3.14 0 5.7 2.56 5.7 5.7s-2.56 5.7-5.7 5.7A5.71 5.71 0 0 1 1.3 8c0-3.14 2.56-5.7 5.7-5.7zM7 1C3.14 1 0 4.14 0 8s3.14 7 7 7 7-3.14 7-7-3.14-7-7-7zm1 3H6v5h2V4zm0 6H6v2h2v-2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/microsoft/vscode-docker/issues/869" target="_blank" rel="noopener nofollow ugc">Support third-party registries</a>
    </h4>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2019-03-27" data-time="22:42:59" data-timezone="UTC">10:42PM - 27 Mar 19 UTC</span>
      </div>


      <div class="user">
        <a href="https://github.com/StephenWeatherford" target="_blank" rel="noopener nofollow ugc">
          <img alt="StephenWeatherford" src="https://user-images.githubusercontent.com/35611074/181096734-c6cd5abf-ab5d-4734-ac7b-bb59ff90900a.png" class="onebox-avatar-inline" width="20" height="20">
          StephenWeatherford
        </a>
      </div>
    </div>

    <div class="labels">
        <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">
          enhancement
        </span>
        <span style="display:inline-block;margin-top:2px;background-color: #B8B8B8;padding: 2px;border-radius: 4px;color: #fff;margin-left: 3px;">
          registries
        </span>
    </div>
  </div>
</div>

  <div class="github-row">
    <p class="github-body-container">Specifically requested (but goal is generic support), in no particular order:

<span class="show-more-container"><a href="" rel="noopener" class="show-more">…</a></span><span class="excerpt hidden">
Open:
- [ ] goharbor.io (see #878, #3449)
- [ ] ECR (Amazon, see #3202)
- [ ] GCR (Google) (+3 see #394, #2890)
- [ ] quay.io (+4 see #390)
- [ ] DTR (Enterprise Docker trusted registry) (see #3325)
- [ ] Cloudsmith (see #1887)
- [ ] Aliyun (see #1964)
- [ ] Portus (see #2735)
- [ ] Digital Ocean (see #2879)

**Blocked** on no/partial Registry V2 support:
- [ ] GitHub (see #1525, #1938)
- [ ] GitLab self-hosted (+5 see #1832, #2457, #2735, #3119, #3271)
- [ ] `harbor.korporate.org` (see #1927)
- [ ] GitLab Groups (see #1970)

Done:
- [x] DockerHub
- [x] Azure
- [x] GitLab (+25-ish see #162)
- [x] Generic V2 registry (+2)</span></p>
  </div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Seems the ball has been dropped, is there an open issue for this backlog feature request?</p>

**Comment by tclose** [comment]
<p>A fix to his issue is would be +1 from me too</p>

**Comment by cpotter302** [comment]
Are there any updates on this? I am currently working on a use case to download blobs from third party registries including ghcr.io

**Comment by mirkobrombin** [comment]
Updates on this?

**Comment by dtrunk90** [comment]
I can't use ghcr.io on my Synology DSM Container Manager as a registry due to this missing feature. I had to run the containers manually through SSH. As a result I'm not getting notified about image updates.

**Comment by anuragagrawal0430** [comment]
Is anyone working on this backlog item?

**Comment by Denney-tech** [comment]
Would love to see an update on this backlog item.

**Comment by BROngineer** [comment]
Any update on this feature?
